### ISSUE TITLE
Activity environment deadline is not updated for activity retries

### ISSUE THREAD

**Issue by shaunco** [temporalio/temporal#1139]
## Expected Behavior
When an activity is retried, the `deadline` value in the activity environment should reflect the start time of the retry attempt + StartToCloseTimeout.

## Actual Behavior
When an activity is retried, the `deadline` value in the activity environment is still the original activity execution time + StartToCloseTimeout.

Nothing changed in the go-sdk (still 1.2.0), and this deadline value comes from Temporal server. Additionally, this worked fine in 1.3.2 but is broken in 1.5.0.

While I can ignore it and just use context.Background(), it seems the deadline in this context should be correct.

## Steps to Reproduce the Problem

  1. While the worker is in a debugger, set a breakpoint at the start of an activity
  1. Have a workflow execute an activity
  1. Terminate the worker
  1. Wait a few minutes (or until StartToCloseTimeout has passed)
  1. Restart the worker in the debugger, with the same activity breakpoint
  1. Note that ctx is immediately past its deadline by checking `ctx.Err()`, which will return `context deadline exceeded`

## Specifications

  - Version: 1.5.0  (was working fine in 1.3.2)
  - Platform: Ubuntu / go-sdk

**Comment by wxing1292** [comment]
@shaunco 

Can you please provide the activity timeout settings?

**Comment by shaunco** [comment]
They are the defaults I copied from https://docs.temporal.io/docs/go-run-your-first-app/#the-workflow-function as I've yet to have a need to make them different.

```go
    // RetryPolicy specifies how to automatically handle retries if an Activity fails.
    retrypolicy := &temporal.RetryPolicy{
        InitialInterval:    time.Second,
        BackoffCoefficient: 2.0,
        MaximumInterval:    time.Minute,
        MaximumAttempts:    500,
    }
    options := workflow.ActivityOptions{
        // Timeout options specify when to automatically timeout Actvitivy functions.
        StartToCloseTimeout: time.Minute,
        // Optionally provide a customized RetryPolicy.
        // Temporal retries failures by default, this is just an example.
        RetryPolicy: retrypolicy,
    }
```

**Comment by wxing1292** [comment]
seems that i cannot repo the issue, let us sync in slack (Wenquan Xing)

activity config
```go
retryPolicy := &temporal.RetryPolicy{
	InitialInterval:    time.Second,
	BackoffCoefficient: 2.0,
	MaximumInterval:    time.Minute,
	MaximumAttempts:    500,
}
activityCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
	TaskQueue:              taskQueueName,
	StartToCloseTimeout:    time.Minute,
	RetryPolicy:            retryPolicy,
})
```

activity function
```go
func retryOnTimeoutActivity(ctx context.Context, scheduledTimeNanos int64) (int, error) {

	fmt.Print("#### ")
	deadline, ok := ctx.Deadline()
	fmt.Printf("%v-%v\n", deadline, ok)
	fmt.Print("#### ")

	return 0, errors.New("123")
}
```

```
#### 2020-12-29 21:32:24.744947 +0000 UTC-true
#### 2020-12-29 21:32:25.754631 +0000 UTC-true
#### 2020-12-29 21:32:27.765819 +0000 UTC-true
#### 2020-12-29 21:32:31.773322 +0000 UTC-true
#### 2020-12-29 21:32:39.783182 +0000 UTC-true
#### 2020-12-29 21:32:54.757266 +0000 UTC-true
#### 2020-12-29 21:32:55.766048 +0000 UTC-true
#### 2020-12-29 21:32:55.793084 +0000 UTC-true
#### 2020-12-29 21:32:57.781462 +0000 UTC-true
#### 2020-12-29 21:33:01.790029 +0000 UTC-true
#### 2020-12-29 21:33:09.800278 +0000 UTC-true
```

**Comment by shaunco** [comment]
On my side the retry executions land in https://github.com/temporalio/sdk-go/blame/56a1a18c4db896e1a8d4fe4dbf95eb97131fc01f/internal/activity.go#L264 ... which was added 19 days ago. It looks like [`task.GetScheduledTime()`](https://github.com/temporalio/sdk-go/blob/56a1a18c4db896e1a8d4fe4dbf95eb97131fc01f/internal/activity.go#L253) is the value that never changes, so even though I never set a `ScheduleToCloseTimeout` it seems to be set to 60 seconds.

Try letting your retry task execute again >60 seconds later to see if you end up in the "minimum of two deadlines" code that was added.

**Comment by wxing1292** [comment]
thanks for reporting: #1141

**Comment by shaunco** [comment]
Awesome! Thank you!

**Comment by wxing1292** [comment]
please use the patch release https://github.com/temporalio/temporal/releases/tag/v1.5.1

**Comment by shaunco** [comment]
Confirmed fixed in 1.5.1

Thanks for the amazingly fast turn around!
